fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java

```diff
@@ -0,0 +1,240 @@
+        if (fieldSetPredicate.test(message)) {
+            return fieldExtractor.apply(message);
+        }
+        throw new RecordCoreException("Field is expected to be set but message does not have field.");
+    }
+
+    /**
+     * Helper method to check for a field's existence in the protobuf message passed in. If the field exists, the method
+     * then extracts and returns the field's value. The method returns {@code null} if the field does not exist.
+     * Note that this method is mostly equivalent to {@code message.getSomeField()} if
+     * {@code someField} does not use a default value, but also works for primitive values.
+     * @param message protobuf message
+     * @param fieldSetPredicate predicate to check if a field is set
+     * @param fieldExtractor extractor extracting the field's value
+     * @param <M> the specific message type
+     * @param <T> the specific field's type
+     * @return the field's value
+     */
+    @Nullable
+    public static <M extends Message, T> T getFieldOrNull(@Nonnull M message,
+                                                          @Nonnull final Predicate<M> fieldSetPredicate,
+                                                          @Nonnull final Function<M, T> fieldExtractor) {
+        if (fieldSetPredicate.test(message)) {
+            return fieldExtractor.apply(message);
+        }
+        return null;
+    }
+
+    /**
+     * {@link BiMap} to cache enum relationships. This helper has been added to avoid using huge switch
+     * statements that relate domain enums and proto enums.
+     * Note that both enum classes must use the same enum value names and ordinals.
+     * @param domainEnumClass domain enum
+     * @param protoEnumClass proto enum
+     * @param <E1> domain enum type parameter
+     * @param <E2> proto enum type parameter
+     * @return a {@link BiMap} that can be used to look up the <em>other</em> enum value for a given enum value
+     */
+    @Nonnull
+    public static <E1 extends Enum<E1>, E2 extends Enum<E2> & ProtocolMessageEnum> BiMap<E1, E2> protoEnumBiMap(@Nonnull final Class<E1> domainEnumClass,
+                                                                                                                @Nonnull final Class<E2> protoEnumClass) {
+        Verify.verify(domainEnumClass.isEnum());
+        Verify.verify(protoEnumClass.isEnum());
+        final E1[] javaEnumConstants = domainEnumClass.getEnumConstants();
+        final E2[] protoEnumConstants = protoEnumClass.getEnumConstants();
+        Verify.verify(javaEnumConstants.length == protoEnumConstants.length);
+
+        final EnumBiMap<E1, E2> enumBiMap = EnumBiMap.create(domainEnumClass, protoEnumClass);
+        for (int i = 0; i < javaEnumConstants.length; i ++) {
+            final E1 e1 = javaEnumConstants[i];
+            final E2 e2 = protoEnumConstants[i];
+            Verify.verify(e1.ordinal() == e2.ordinal());
```

Comment:
Hm, do we care if the ordinals match? I can think of examples where we may have trouble keeping the ordinals aligned (e.g., if an enum value is added in the middle of a POJO definition because it groups well with existing values, but we want to put it at the end of the proto definition to simplify enum number assignment).

I'm okay with keeping this check here for now and then removing it later if/when we find it to be problematic